### ISSUE TITLE
How does one construct form parts with a particular content-type?

### ISSUE THREAD

**Issue by shonfeder** [dinosaure/multipart_form#33]
Hello!

I'm trying to make use of this library along side Cohttp.

I'm having trouble with constructing `parts`. In particular, I need to construct parts with type `Content-Type: application/octet-stream` or `Content-Type: text/plain; charset=utf8`, but afaict, the current API leaves no way to set a content type other than what is supplied by a given `~header` labeled argument to `part`, and it seems there is no way to construct a value of `Header.t` other than parsing it out. Am i misunderstanding, or is this a basic limitation of the current API?

Thanks in advance for any pointers!

**Comment by dinosaure** [comment]
> the current API leaves no way to set a content type other than what is supplied by a given ~header labeled argument to part, and it seems there is no way to construct a value of Header.t other than parsing it out.

You are totally true, the `multipart_form` encoder is pretty _new_ and not really used currently. That's why there are lacks in the API that I haven't really noticed. I will try to propose something as soon as I can but if you have some idea about a better API to generate a `multipart_form` stream, please tell me and I will integrate them :+1:.

**Comment by dinosaure** [comment]
See #34 for beginning.

**Comment by dinosaure** [comment]
@shonfeder any feedbacks?

**Comment by shonfeder** [comment]
I ended up needing to hack together a very simplistic and naive (and
non-streaming) abstraction for generating multipart form data in order to
unblock my project. It is a much simpler interface, but of course has none of
the nice features of this library.  Here's the interface, in case its any use to
you: 

```ocaml
module Part : sig
  type t
  (** The type of multipart-form parts *)

  val v : name:string -> ?filename:string -> ?typ:string -> string -> t
  (** [v ~name content] is a part with the given [name] containing
      supplied [content].

      Optional parameters

      - [filename]: Specify a filename for the part content
      - [typ]: Specify the content type for the part. This should be a
        valid MIME type. *)

  val add_to_buffer : Buffer.t -> t -> unit
  val to_string : t -> string
end

type t
(** The type of multipart forms *)

val v : ?boundary:string -> Part.t list -> t
(** [v parts] is a new multipart-form composed of the [parts].

    Optionally, a custom [boundary] marker can be supplied. Otherwise,
    a boundary is generated as a UUID. *)

val to_string : t -> string

val add_header : t -> Cohttp.Header.t -> Cohttp.Header.t
(** [add_header form headers] is [headers] with the appropriate
    multipart/form-data content type declaration added *)
````

Aside from the limits on constructing the parts, the other substantial hurdle I
hit with your (very nice!) library was trying to integrate it with
`Cohttp_lwt.Body` via  `Cohttp_lwt.Body.of_stream`. I didn't end up figuring out
how to preserve the streaming when constructing a request.

Sorry I don't have any bright ideas on the interface here, but thank you for your great work!

**Comment by dinosaure** [comment]
@shonfeder

Thank you for your feedback. I have decided to offer a helper for CoHTTP with a new package. The idea is still to create the "multipart" value on one side. Then a helper can send this value with the CoHTTP API. Finally, a helper exists for the server to handle a multipart/form request. It would be very appreciated if you could test this API to validate its operation (there may be bugs notably because of the "\r\n" but if this is the case, I can make a patch to fix this specific issue).

You can test it via:
```sh
$ opam pin add -y git+https://github.com/dinosaure/multipart_form.git#header-inj
```

I don't have much time to really propose something better but I'm always available to improve the library.

**Comment by shonfeder** [comment]
Thanks, @dinosaure! I will give this a try when I have time. Hopefully this week.

**Comment by shonfeder** [comment]
Hi again! The package for interfacing to Cohttp looks good to me so far, tho I havent run against my end to end tests yet.

The last snag I am hitting is figuring oup how to stream in the content of a file *from within the context of Lwt*. If i could just use `In_channel` it would be straight forward, but producing a `(string * int 8

**Comment by dinosaure** [comment]
> The last snag I am hitting is figuring out how to stream in the content of a
file from within the context of Lwt. If I could just use In_channel it would
be straight forward, but producing a unit -> (string * int * int) option that
reads from a file opened via Lwt_io is proving more difficult than I'd anticipated
difficult.

Currently, there is no way to produce a `Multipart_form.stream` from a `lwt` context (in other words, going from a `string Lwt_stream.t` to `unit -> string`). The problem can be solved (if one wishes to keep the `multipart_form` agnostic side) with higher kinded polymorphism but it should be:
1) the subject of another PR
2) would complicate the project with a not-so-obvious API (I've already experimented with this on some of [my libraries](https://github.com/mirage/ocaml-dkim) and the interest remains quite limited)

The immediate solution would be not to use `lwt` to generate the stream as for example:
```ocaml
let stream_of_file file =
  let ic = open_in file in
  let tp = Bytes.create 0x1000 in
  fun () -> match input ic tp 0 0x1000 with
  | 0 -> close_in ic ; None
  | len -> Some (Bytes.substring tp 0 len)
  | exception End_of_file -> close_in ic ; None
```

However, if there is a real need for its use with `lwt`, I can indeed make an effort and propose in the next weeks something with high-kind polymorphism but again, this is only if I see a real interest.

**Comment by shonfeder** [comment]
The new package worked perfectly for us. You can see an example of our usage at https://github.com/XFFS/oopenai/pull/12/files#diff-1eaadd9ced8dc993e6b81eb3d2b117373a247488b85095a38ac61caf0acc2987R80

Let me know if there's anything I can do to support you in releasing the new package adition!

**Comment by dinosaure** [comment]
Thanks, I think I can close this issue now. I will cut a release now.

**Comment by shonfeder** [comment]
Awesome. Thanks very much! :pray:
